### PR TITLE
Allow adding new members to a collection through a POST request

### PR DESCRIPTION
**Pull request by KanishManuja-MS**
### Issues

*This pull request adds a new feature which allows adding new members to a collection of complex types, enums, or primitive types through a POST request.  *

### Description

This pull request defines the property routing convention for a POST request on a property and adds end to end testcases for the above-mentioned feature.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

Feature documentation needs to be added.

### CODE REVIEW

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs [hunk 1]
```diff
@@ -184,12 +185,48 @@ public void Read_Primitive(object obj, string edmType, string value)
...
             settings.SetContentType(ODataFormat.Json);
 
+            Type type = obj == null ? typeof(int) : obj.GetType();
+
             ODataMessageWriter messageWriter = new ODataMessageWriter(message as IODataResponseMessage, settings, model);
             ODataMessageReader messageReader = new ODataMessageReader(message as IODataResponseMessage, new ODataMessageReaderSettings(), model);
             ODataSerializerContext writeContext = new ODataSerializerContext { RootElementName = "Property", Model = model };
-            ODataDeserializerContext readContext = new ODataDeserializerContext { Model = model };
+            ODataDeserializerContext readContext = new ODataDeserializerContext { Model = model, ResourceType = type };
+
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: remove line

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs [hunk 2]
```diff
@@ -184,12 +185,48 @@ public void Read_Primitive(object obj, string edmType, string value)
...
+            Assert.NotNull(edmType);
+            Assert.NotNull(value);
+            Assert.Equal(obj, deserializer.Read(messageReader, type, readContext));
+        }
+
+        [Theory]
+        [MemberData(nameof(EdmPrimitiveData))]
+        public void Read_Primitive(object obj, string edmType, string value)
+        {
+                    // Arrange
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: make the code align.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 3]
```diff
@@ -539,6 +598,49 @@ public async Task PutCollectionComplexTypeProperty(string modelMode)
...
 
+        [Theory]
+        [MemberData(nameof(NewCollectionMembers))]
+        // POST ~/Windows(3)/OptionalShapes
+        public async Task PostToCollectionComplexTypeProperty(string modelMode, string jObject, string targetPropertyResource)
+        {
+            //Arrange
+            string serviceRootUri = string.Format("{0}/{1}", BaseAddress, modelMode).ToLower();
+            string requestUri = serviceRootUri + "/Windows(3)/"+ targetPropertyResource;
+            int count = 0;
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: It is used in the second assertion.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs [hunk 4]
```diff
@@ -245,6 +256,35 @@ public ITestActionResult ReplaceOptionalShapes(int key, IEnumerable<Shape> shape
...
+        [HttpPost]
+        public ITestActionResult PostToOptionalShapes(int key, [FromBody]Shape newShape)
+        {
+            Window window = _windows.FirstOrDefault(w => w.Id == key);
+            if (window == null)
+            {
+                return NotFound();
+            }
+            window.OptionalShapes.Add(newShape);
+            return Ok(window.OptionalShapes);
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: As per the email discussion, I changed this to Updated.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs [hunk 5]
```diff
@@ -414,6 +415,35 @@ public async Task AddEntity()
             }
         }
 
+
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs]: Done.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs [hunk 6]
```diff
@@ -245,6 +256,35 @@ public ITestActionResult ReplaceOptionalShapes(int key, IEnumerable<Shape> shape
...
+        }
+
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: removed.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataDeserializationTestsCommon.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            return new ODataMessageReader(oDataRequestMessage, new ODataMessageReaderSettings(), edmModel);
+        }
+
+        internal static IODataRequestMessage GetODataMessage(string content)
+        {
+            // While NetCore does not use this for AspNet, it can be used here to create
+            // an HttpRequestODataMessage, which is a Test type that implments IODataRequestMessage
+            // wrapped around an HttpRequestMessage.
+            HttpRequestMessage request = new HttpRequestMessage(new HttpMethod("Patch"), "http://localhost/OData/Suppliers(1)/Address");
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataDeserializationTestsCommon.cs]: Fixed

File: src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs [hunk 8]
```diff
@@ -66,8 +66,13 @@ public virtual object ReadPrimitive(ODataProperty primitiveProperty, ODataDeseri
...
                 throw Error.ArgumentNull("primitiveProperty");
             }
+            //Try and change the value appropriately if type is specified
+            if (readContext != null && readContext.ResourceType != null && primitiveProperty.Value != null)
+            {
+                return EdmPrimitiveHelpers.ConvertPrimitiveValue(primitiveProperty.Value, readContext.ResourceType);
+            }
 
-            return primitiveProperty.Value;
+            return primitiveProperty.Value;
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs]: >  [](start = 43, length = 1)

remove it

File: src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs [hunk 9]
```diff
@@ -66,8 +66,13 @@ public virtual object ReadPrimitive(ODataProperty primitiveProperty, ODataDeseri
             {
                 throw Error.ArgumentNull("primitiveProperty");
             }
+            //Try and change the value appropriately if type is specified
+            if (readContext != null && readContext.ResourceType != null && primitiveProperty.Value != null)
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs]: >readContext != null [](start = 16, length = 19)

Do we allow "readContext== null" ?

if not, maybe it's better to add the argument check, like the "PrimitiveProperty"

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xuzhg) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: Would you check Updated should accept the "window", not the "window.OptionalShapes" ?

---
In reply to: [222784262](https://github.com/OData/webapi/pull/1628#discussion_r222784262) [](ancestors = 222784262)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 10]
```diff
@@ -48,6 +48,116 @@ public ComplexTypeInheritanceTests(WebHostTestFixture fixture)
             }
         }
 
+        public static TheoryDataSet<string, string, string,bool> NewCollectionMembers
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: >NewCollectionMembers [](start = 65, length = 20)

rename it meaningful

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 11]
```diff
@@ -48,6 +48,116 @@ public ComplexTypeInheritanceTests(WebHostTestFixture fixture)
...
+
+                   { "convention", @"
+{
+        '@odata.type':'#Microsoft.Test.E2E.AspNet.OData.ComplexTypeInheritance.Polygon',
+        'HasBorder':true,'Vertexes':[
+            {'X':21,'Y':12},
+            {'X':32,'Y':23},
+            {'X':14,'Y':41}
+        ]
+}",
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: you can combine the tests using codes. See "MediaTypes" foreach (...) clause.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 12]
```diff
@@ -539,6 +649,63 @@ public async Task PutCollectionComplexTypeProperty(string modelMode)
...
+            string requestUri = serviceRootUri + "/Windows(3)/"+ targetPropertyResource;
+            int count = 0;
+
+            //Setup the post request
+            var requestForPost = new HttpRequestMessage(HttpMethod.Post, requestUri);
+            requestForPost.Content = new StringContent(content:jObject, encoding: Encoding.UTF8, mediaType: "application/json");
+            if (returnRepresentation)
+            {
+                requestForPost.Headers.Add("Prefer", "return=representation");
+            }
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Move these codes after the "Get request"

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 13]
```diff
@@ -539,6 +649,63 @@ public async Task PutCollectionComplexTypeProperty(string modelMode)
...
+                var json = await getResponse.Content.ReadAsObject<JObject>();
+                var state=json.GetValue("value") as JArray;
+                count = state.Count;
+            }
+
+            //Act & Assert
+            HttpResponseMessage response = await Client.SendAsync(requestForPost);
+            string contentOfString = await response.Content.ReadAsStringAsync();
+
+
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: remove unused whitespace line

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeController.cs [hunk 14]
```diff
@@ -123,7 +123,7 @@ public ITestActionResult GetFilesModifiedAt(DateTimeOffset modifiedDate)
 
             return Ok(files);
         }
-
+
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeController.cs]: remove the whitespace

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeController.cs [hunk 15]
```diff
@@ -138,6 +138,36 @@ public ITestActionResult CopyFiles(int key, [FromBody]ODataActionParameters para
...
+            try
+            {
+                file.ModifiedDates.Add(newDateTime);
+
+            }
+            catch (Exception e)
+            {
+                string s = e.HelpLink;
+            }
+            return Updated(newDateTime);
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeController.cs]: >newDateTime [](start = 27, length = 11)

it should be "file"?

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs [hunk 16]
```diff
@@ -235,5 +235,69 @@ public void SelectAction_ReturnsNull_IfPatchToCollectionProperty()
...
+
+            // Act
+            string selectedAction = SelectActionHelper.SelectAction(new PropertyRoutingConvention(), odataPath, request, actionMap);
+
+            // Assert
+            Assert.Null(selectedAction);
+            Assert.Empty(SelectActionHelper.GetRouteData(request).Values);
+        }
+
+
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs]: remove

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 10 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Does PostToCollectionNewComplextTypeMembers seem good?

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: Yes, as per Mike's discussion, return=representation should return the target collection.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs (hunk 2 above)
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: fixed

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs (hunk 1 above)
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: fixed

File: src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs (hunk 8 above)
Comment (KanishManuja-MS) [reply src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs]: fixed

File: src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs (hunk 9 above)
Comment (KanishManuja-MS) [reply src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataPrimitiveDeserializer.cs]: Added the readContext null check.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 11 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Done.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 12 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Done

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeController.cs (hunk 15 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeController.cs]: It should be file.ModifiedDates. Updated.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs (hunk 16 above)
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs]: removed whitespace

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs [hunk 17]
```diff
@@ -414,6 +415,35 @@ public async Task AddEntity()
...
+            await ResetDatasource();
+            //Arrange
+            string requestUri = this.BaseAddress + "/convention/Employees/2/SkillSet?$format=application/json;odata.metadata=none";
+            var requestForPost = new HttpRequestMessage(HttpMethod.Post, requestUri);
+
+            requestForPost.Content = new StringContent(content: @"{
+                    'value':'Sql'
+                    }", encoding: Encoding.UTF8, mediaType: "application/json");
+           
+            using (HttpResponseMessage response = await this.Client.SendAsync(requestForPost))
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs]: fixed

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs [hunk 18]
```diff
@@ -414,6 +415,35 @@ public async Task AddEntity()
             }
         }
 
+
+
+        [Fact]
+        public async Task PostToEnumCollection()
+        {
+            await ResetDatasource();
+            //Arrange
```
[the remark this answers is not in the document]
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs]: fixed

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 19]
```diff
@@ -539,6 +585,62 @@ public async Task PutCollectionComplexTypeProperty(string modelMode)
                 contentOfString));
         }
 
+        [Theory]
+        [MemberData(nameof(PostToCollectionNewComplextTypeMembers))]
+        // POST ~/Windows(3)/OptionalShapes
+        public async Task PostToCollectionComplexTypeProperty(string modelMode, string jObject, string targetPropertyResource, bool returnRepresentation)
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: rename it

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs [hunk 20]
```diff
@@ -83,6 +84,35 @@ public async Task ModelBuilderTest(string modelMode)
             Assert.False(modifiedDates.Type.IsNullable);
         }
 
+        [Theory]
+        [InlineData("convention")]
+        [InlineData("explicit")]
+        public async Task PostToDateTimeCollection(string modelMode)
+        {
+            await ResetDatasource();
+            //Arrange
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs]: >//Arrange [](start = 12, length = 9)

If you want to add // Arrange, please add // Act // Assert as well.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs [hunk 21]
```diff
@@ -83,6 +84,35 @@ public async Task ModelBuilderTest(string modelMode)
...
+            }
+
+            using (HttpResponseMessage response = await this.Client.GetAsync(requestUri))
+            {
+                response.EnsureSuccessStatusCode();
+
+                var json = await response.Content.ReadAsObject<JObject>();
+                var result = json.GetValue("value") as JArray;
+                Assert.Equal(4,result.Count);
+            }
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs]: if you want to test the count, i think you can use "$count" segment. For example:  ~/Fiels/1/ModifiedDates/$count

Comment (xuzhg) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs]: Add the count check before the post as well.

---
In reply to: [223518830](https://github.com/OData/webapi/pull/1628#discussion_r223518830) [](ancestors = 223518830)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 22]
```diff
@@ -539,6 +585,62 @@ public async Task PutCollectionComplexTypeProperty(string modelMode)
...
+            }
+
+            //Act & Assert
+            HttpResponseMessage response = await Client.SendAsync(requestForPost);
+            string contentOfString = await response.Content.ReadAsStringAsync();
+
+            if(returnRepresentation)
+            {
+                JObject contentOfJObject = await response.Content.ReadAsObject<JObject>();
+                var result = contentOfJObject.GetValue("value") as JArray;
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: >result [](start = 20, length = 6)

It looks only the count needed, you can use "$count" segment to get the count directly.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs [hunk 23]
```diff
@@ -414,6 +415,50 @@ public async Task AddEntity()
...
+            string requestUri = this.BaseAddress + "/convention/Employees/2/SkillSet?$format=application/json;odata.metadata=none";
+            //Get the count before the post
+            int count = 0;
+            using (HttpResponseMessage response = await this.Client.GetAsync(requestUri))
+            {
+                response.EnsureSuccessStatusCode();
+
+                var json = await response.Content.ReadAsObject<JObject>();
+                var result = json.GetValue("value") as JArray;
+                count = result.Count;
```
Comment (xuzhg) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs]: >count [](start = 16, length = 5)

Same as here if you only want to get the count using "$count" segment.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs [hunk 24]
```diff
@@ -199,6 +236,93 @@ public void Read_Primitive(object obj, string edmType, string value)
...
+            settings.SetContentType(ODataFormat.Json);
+
+            Type type = obj == null ? typeof(int) : expected.GetType();
+
+            ODataMessageWriter messageWriter = new ODataMessageWriter(message as IODataResponseMessage, settings, model);
+            ODataMessageReader messageReader = new ODataMessageReader(message as IODataResponseMessage, new ODataMessageReaderSettings(), model);
+            ODataSerializerContext writeContext = new ODataSerializerContext { RootElementName = "Property", Model = model };
+            ODataDeserializerContext readContext = new ODataDeserializerContext { Model = model, ResourceType = type };
+
+
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: remove

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs [hunk 25]
```diff
@@ -199,6 +236,93 @@ public void Read_Primitive(object obj, string edmType, string value)
...
+            
+            serializer.WriteObject(obj, type, messageWriter, writeContext);
+            stream.Seek(0, SeekOrigin.Begin);
+
+            // Act && Assert
+            Assert.Equal(expected, deserializer.Read(messageReader, type, readContext));
+        }
+
+
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: remove

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs [hunk 26]
```diff
@@ -199,6 +236,93 @@ public void Read_Primitive(object obj, string edmType, string value)
...
+            
+            // Act
+            object value = deserializer.Read(ODataDeserializationTestsCommon.GetODataMessageReader(ODataDeserializationTestsCommon.GetODataMessage(content, new HttpRequestMessage(new HttpMethod("Patch"), "http://localhost/OData/Suppliers(1)/Address")), model), type, readContext);
+
+            // Assert
+            Assert.Equal(expected,value);
+        }
+
+        [Fact]
+        public void ReadFromStreamAsync_RawGuid()
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: >ReadFromStreamAsync_RawGuid [](start = 20, length = 27)

Guid can't test in above "Test method"?, like other types (single, double)?

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataDeserializationTestsCommon.cs [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+using System.Net.Http;
+using System.Net.Http.Headers;
+using Microsoft.OData;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Test.Formatter.Deserialization
+{
+    internal class ODataDeserializationTestsCommon
+    {
+
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataDeserializationTestsCommon.cs]: You don't need to create a new file to hold these two methods, you can leave them in the ODataEnumDescrializerTests.cs and make them public static method.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs [hunk 28]
```diff
@@ -235,5 +235,67 @@ public void SelectAction_ReturnsNull_IfPatchToCollectionProperty()
...
+            // Assert
+            Assert.NotNull(selectedAction);
+            Assert.Equal(prefix + "OtherAccountsFromCustomer", selectedAction);
+            Assert.Single(SelectActionHelper.GetRouteData(request).Values);
+            Assert.Equal(7, SelectActionHelper.GetRouteData(request).Values["key"]);
+        }
+
+        [Theory]
+        [InlineData("Get", "Get")]
+        [InlineData("Put", "PutTo")]
```
Comment (xuzhg) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs]: >"Put", "PutTo")] [](start = 20, length = 16)

Why "Complex" test case has "Delete", enum doesn't?

**Comment by xuzhg** [comment]
I leave some comments related to the test cases. Other It looks fine to me.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs [hunk 29]
```diff
@@ -539,6 +585,62 @@ public async Task PutCollectionComplexTypeProperty(string modelMode)
                 contentOfString));
         }
 
+        [Theory]
+        [MemberData(nameof(PostToCollectionNewComplextTypeMembers))]
```
Comment (mikepizzo) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: >t [](start = 53, length = 1)

nit: extra "t"

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 22 above)
Comment (mikepizzo) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: Yes, he could make a separate GET request to /$count, but here he's specifically checking the result of the POST request.

---
In reply to: [223519003](https://github.com/OData/WebApi/pull/1628#discussion_r223519003) [](ancestors = 223519003)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs (hunk 21 above)
Comment (mikepizzo) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs]: As above; he's checking the count of the response from the POST request to make sure it includes the added member, which is good.

---
In reply to: [223518830](https://github.com/OData/WebApi/pull/1628#discussion_r223518830) [](ancestors = 223518830)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs (hunk 23 above)
Comment (mikepizzo) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Enums/EnumsTest.cs]: As above; this is validating the response from the POST, not a separate GET.

---
In reply to: [223519142](https://github.com/OData/WebApi/pull/1628#discussion_r223519142) [](ancestors = 223519142)

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs (hunk 28 above)
Comment (mikepizzo) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs]: +1: I assume we support Delete here too?

---
In reply to: [223519938](https://github.com/OData/WebApi/pull/1628#discussion_r223519938) [](ancestors = 223519938)

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikepizzo) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceControllers.cs]: thumbs up!

---
In reply to: [223472984](https://github.com/OData/WebApi/pull/1628#discussion_r223472984) [](ancestors = 223472984)

**Comment by mikepizzo** [comment]
Looks good to me too, once you clean up some minor test issues.

---
In reply to: [428006170](https://github.com/OData/WebApi/pull/1628#issuecomment-428006170) [](ancestors = 428006170)

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs (hunk 26 above)
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataPrimitiveDeserializerTests.cs]: Guid is different compared to other simple primitive types as new Guid() or Guid.NewGuid is not treated as a constant.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs (hunk 28 above)
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/Conventions/PropertyRoutingConventionTests.cs]: EnumCollections do support delete but the property needs to be nullable for that, which is not the case with this property.

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataDeserializationTestsCommon.cs (hunk 27 above)
Comment (KanishManuja-MS) [reply test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Formatter/Deserialization/ODataDeserializationTestsCommon.cs]: That would have been the simpler thing to do but referencing ODataEnumDeserializer in ODataPrimitiveDeserializer code did not seem the right thing to do.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs (hunk 20 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs]: Adding Act & Assert as they are coupled tightly.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs (hunk 22 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ComplexTypeInheritance/ComplexTypeInheritanceTests.cs]: I agree that I could make the first request to get the count. With that being said, I do not see a major difference for my purpose and I felt doing a get was simpler.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs (hunk 21 above)
Comment (KanishManuja-MS) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DateTimeSupport/DateTimeTest.cs]: Added a request for the pre-condition.

**Pull request merged**